### PR TITLE
[codegen/go] Call site defaults for Pulumi Object types

### LINKED ISSUE #8132: [codegen/nodejs,go] Fail to emit defaults for plain objects
## Hello!

- Vote on this issue by adding a 👍 reaction
- To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already)

## Issue details

The C# and Python SDKs generate helper code to check configured environment variables for Providers. It appears that the Go and TypeScript SDKs both have a bug that prevents this code from being generated on complex Provider types.

### Steps to reproduce

Here's what the schema looks like:

```
"provider": {
    "description": "The provider type for the kubernetes package.",
    "type": "object",
    "inputProperties": {
    ...
        "helmReleaseSettings": {
            "$ref": "#/types/kubernetes:index:HelmReleaseSettings",
            "description": "BETA FEATURE - Options to configure the Helm Release resource."
        },
    ...
},
"types": {
  "kubernetes:index:HelmReleaseSettings": {
      "description": "BETA FEATURE - Options to configure the Helm Release resource.",
      "properties": {
      ...
          "driver": {
              "type": "string",
              "description": "The backend storage driver for Helm. Values are: configmap, secret, memory, sql.",
              "default": "secret",
              "defaultInfo": {
                  "environment": [
                      "PULUMI_K8S_HELM_DRIVER"
                  ]
              }
          },
          "pluginsPath": {
              "type": "string",
              "description": "The path to the helm plugins directory.",
              "defaultInfo": {
                  "environment": [
                      "PULUMI_K8S_HELM_PLUGINS_PATH"
                  ]
              }
          },
      ...
      "type": "object"
  },
},
```

Here's an example of the helpers generated for Python:
```py3
@pulumi.input_type
class HelmReleaseSettingsArgs:
    def __init__(__self__, *,
                 driver: Optional[pulumi.Input[str]] = None,
                 plugins_path: Optional[pulumi.Input[str]] = None,
                 registry_config_path: Optional[pulumi.Input[str]] = None,
                 repository_cache: Optional[pulumi.Input[str]] = None,
                 repository_config_path: Optional[pulumi.Input[str]] = None,
                 suppress_beta_warning: Optional[pulumi.Input[bool]] = None):
        """
        BETA FEATURE - Options to configure the Helm Release resource.
        :param pulumi.Input[str] driver: The backend storage driver for Helm. Values are: configmap, secret, memory, sql.
        :param pulumi.Input[str] plugins_path: The path to the helm plugins directory.
        :param pulumi.Input[str] registry_config_path: The path to the registry config file.
        :param pulumi.Input[str] repository_cache: The path to the file containing cached repository indexes.
        :param pulumi.Input[str] repository_config_path: The path to the file containing repository names and URLs.
        :param pulumi.Input[bool] suppress_beta_warning: While Helm Release provider is in beta, by default 'pulumi up' will log a warning if the resource is used. If present and set to "true", this warning is omitted.
        """
        if driver is None:
            driver = _utilities.get_env('PULUMI_K8S_HELM_DRIVER')
        if driver is not None:
            pulumi.set(__self__, "driver", driver)
        if plugins_path is None:
            plugins_path = _utilities.get_env('PULUMI_K8S_HELM_PLUGINS_PATH')
        if plugins_path is not None:
            pulumi.set(__self__, "plugins_path", plugins_path)
    ...
```

Expected: 

Each SDK should generate helper methods that read the configured environment variable as a default.

Actual: 

Only Python and C# generate helpers for complex types.

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

The Go half of #8400.
Fixes #8132
Fixes https://github.com/pulumi/pulumi/issues/7811

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/go/gen.go [hunk 1]
```diff
@@ -1119,6 +1150,78 @@ func (pkg *pkgContext) genPlainType(w io.Writer, name, comment, deprecationMessa
 	fmt.Fprintf(w, "}\n\n")
 }
 
+// Returns the default value used to perform "is default" checks.
+func (pkg *pkgContext) defaultComparisonValue(t schema.Type) string {
```
Comment (t0yv0) [inline pkg/codegen/go/gen.go:1154]: Are you looking for some variation of this? Note that the reflect runtime version is easier to unit-test out.

```
package main

import (
	"fmt"
	"reflect"
)

func isZero(v interface{}) bool {
	if v == nil {
		return true
	}
	return reflect.ValueOf(v).IsZero()
}

func check(x interface{}) {
	fmt.Printf("isZero(%v) == %v\n", x, isZero(x))
}

type foo struct {
	a int
	b *foo
}

type bar interface {
	foo()
}

func main() {
	check(1)
	check(0)
	check("")
	check("a")
	check(foo{})
	check(&foo{}) // NOTE: this is false, as the pointer is non-zero; can code a special case for this
	var b bar
	check(b)
}
```

File: pkg/codegen/go/gen.go [hunk 2]
```diff
@@ -1119,6 +1150,78 @@ func (pkg *pkgContext) genPlainType(w io.Writer, name, comment, deprecationMessa
...
+	switch {
+	case schema.IsPrimitiveType(t):
+		return primitiveNilValue(t)
+	case isNilType(t):
+		return "nil"
+	default:
+		return strings.TrimSuffix(pkg.typeString(t), "Args") + "{}"
```
Comment (t0yv0) [inline pkg/codegen/go/gen.go:1161]: The fact that we use `strings.TrimSuffix` here is a very unfortunate historical artifact, it certainly will be a stumbling block for reviewing this later on... Another argument in favor of reflect version.

File: pkg/codegen/go/gen.go [hunk 3]
```diff
@@ -1310,6 +1413,14 @@ func (pkg *pkgContext) getDefaultValue(dv *schema.DefaultValue, t schema.Type) (
 			return "", err
 		}
 		val = v
+		switch t.(type) {
+		case *schema.EnumType:
+			typeName := strings.TrimSuffix(strings.TrimSuffix(pkg.typeString(t), "Input"), "Ptr")
```
Comment (t0yv0) [inline pkg/codegen/go/gen.go:1418]: It feels like there is some deficiency in typeString here. Perhaps the unwrapping of Input(T) -> T and *T -> T can happen as a transform of `t`, not the string, before we pass it to `pkg.typeString()` and pkg.typeString could take care of `pulumi.Any` as well?

**Comment by t0yv0** [comment]
Got distracted a bit, continuing to review tomorrow morning.

File: pkg/codegen/internal/test/testdata/plain-object-disable-defaults/go/example/funcWithAllOptionalInputs.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	A *HelmReleaseSettings `pulumi:"a"`
+	// Property B
+	B *string `pulumi:"b"`
+}
+
+type FuncWithAllOptionalInputsResult struct {
+	R string `pulumi:"r"`
+}
+
+func FuncWithAllOptionalInputsOutput(ctx *pulumi.Context, args FuncWithAllOptionalInputsOutputArgs, opts ...pulumi.InvokeOption) FuncWithAllOptionalInputsResultOutput {
```
Comment (t0yv0) [inline pkg/codegen/internal/test/testdata/plain-object-disable-defaults/go/example/funcWithAllOptionalInputs.go:34]: There's no defaults here correct? Defaults don't apply to Invokes? Is this by design or deferred for later? Why is  funcWithAllOptionalInputs included in this test suite? I remember authoring it to test the Output version generation.

File: pkg/codegen/nodejs/gen.go [hunk 5]
```diff
@@ -499,38 +499,14 @@ func provideDefaultsFuncNameFromName(typeName string) string {
 // `type` is the type which the function applies to.
 // `input` indicates whither `type` is an input type.
 func (mod *modContext) provideDefaultsFuncName(typ schema.Type, input bool) string {
-	if !isProvideDefaultsFuncRequired(typ) {
+	if !codegen.IsProvideDefaultsFuncRequired(typ) {
```
Comment (t0yv0) [inline pkg/codegen/nodejs/gen.go:502]: This is a no-op change correct? just moving it out to reuse?

**Review by t0yv0 — APPROVED** [review]
Had a few questions but it's LGTM. The codegen code is a bit difficult to follow from the perspective of trying to make sure there's never bugs in it - but that is a problem with how Go codegen is structured, not the current PR, the current PR fits the style, I think we can ship this on the strength of the downstream checks compiling.

One exercise worth trying is maybe running the generated code on some realistic example, just to make sure it works out. I don't see any unit tests added that exercise it. Whenever I modified codegen I routinely found some issues that passed compile checks but would crop up at runtime.

File: pkg/codegen/nodejs/gen.go (hunk 5 above)
Comment (iwahbe) [reply pkg/codegen/nodejs/gen.go:502]: Yep. I just refactored so the code could be shared between Go and TypeScript codegen.

**Comment by iwahbe** [comment]
> One exercise worth trying is maybe running the generated code on some realistic example, just to make sure it works out. I don't see any unit tests added that exercise it. Whenever I modified codegen I routinely found some issues that passed compile checks but would crop up at runtime.

I am relatively confident that this works. I have run "manual" tests, and they work. I have also piggybacked off of other tests that already exist in the codebase. Since you mention it, I think there should be a some unit tests specifically for this. I will add them to the PR.

**Comment by t0yv0** [comment]
> . I have run "manual" tests, and they work. I have also piggybacked off of other tests that already exist in the codebase. Since you mention it, I think there should be a some unit tests specifically for this. I will add them to the PR.

I'm happy to merge based on doing the manual tests, thanks for that!

If you do want to pursue adding a more permanent test by dropping a file in `go-extras` in one of the examples that's great too.

File: pkg/codegen/internal/test/testdata/plain-object-defaults/go-extras/tests/codegen_test.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		RequiredArg: "This is required",
+	}
+	withDefaults := kcs.Defaults()
+	assert.Equal(t, kcs.RequiredArg, withDefaults.RequiredArg)
+	assert.Equal(t, kcs.PluginsPath, withDefaults.PluginsPath)
+	assert.Nil(t, kcs.Driver)
+	assert.Equal(t, withDefaults.Driver, &defaultDriver)
+}
+
+func TestTransitiveObjectDefaults(t *testing.T) {
```
Comment (t0yv0) [inline pkg/codegen/internal/test/testdata/plain-object-defaults/go-extras/tests/codegen_test.go:70]: 👍

**Comment by t0yv0** [comment]
Love the unit tests :chefkiss:

**Pull request merged**